### ISSUE TITLE
Make the Deferred returned by twisted.web.xmlrpc.Proxy.callRemote cancellable

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4377]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)| @jdavisp3 reported|
|-|-|
|Trac ID|trac#4377|
|Type|enhancement|
|Created|2010-03-23 05:49:12Z|

I don't actually have a need to cancel an xmlrpc client request, but I'm trying to understand the new deferred cancel() mechanism. 

Supposing you'd want to allow clients to cancel an xml-rpc request by closing the connection, is this the way to go about it?

Attachments:

* [patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/50a/50a022a513174df7876ffd39e82aaf5c4897146f/d75b3b528276d7a9f30a04b8bccaf74e1d61f67a) (2785 bytes) - added by davep on 2010-03-23 05:49:51Z - 
* [patch.2](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/50a/50a022a513174df7876ffd39e82aaf5c4897146f/9a93ae6b956d75af6dc84d7da45aaab93eab6f2b.2) (4091 bytes) - added by davep on 2010-03-25 02:48:08Z - 
* [patch.3](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/50a/50a022a513174df7876ffd39e82aaf5c4897146f/7514d0275c60358d9c6a276d5f5851b8049052cf.3) (4086 bytes) - added by davep on 2010-03-25 02:54:32Z - 
* [patch.4](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/50a/50a022a513174df7876ffd39e82aaf5c4897146f/6eee34a396de969ca2d8419aea23fbbc9c032a7c.4) (4351 bytes) - added by davep on 2010-04-04 19:14:32Z - Added docstring to explain cancel behavior.
* [patch.5](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/50a/50a022a513174df7876ffd39e82aaf5c4897146f/48d9a149e6712a1dfbefae052b35659bd43eb615.5) (5583 bytes) - added by davep on 2010-04-10 18:34:34Z - Address some review comments.

<details><summary>Searchable metadata</summary>

```
trac-id__4377 4377
type__enhancement enhancement
reporter__davep davep
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1269323352000000 1269323352000000
changetime__1270926771000000 1270926771000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This looks like it's basically the right idea.  A couple specific things about it that jump out at me though:

  * The test should actually verify that the connection is torn down when the Deferred is cancelled.
  * `UserError` isn't what I would have expected this Deferred to fail with in this case.  I haven't run the code, but I assume the test passes as it is written now?  Perhaps this is because connector.disconnect synchronously calls connectionLost on the http protocol instance, which results in the Deferred being errback'd synchronously.  If the Deferred is not errback'd by the canceller function synchronously, then the Deferred cancellation implementation will take responsibility for firing the Deferred with `CancelledError`.

It's something of a judgement call (at least at this point in time :) whether `CancelledError` or `UserError` is more useful to the end user of this API.  I think `CancelledError` is more useful, since you might be interested in the fact that the call failed because someone _told_ it to fail.  `UserError` might indicate a failure due to some network issue and, for example, you might want to automatically retry when getting it.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_2"></a>|@jdavisp3 commented|
|-|-|

Thanks for replying!

I'm attaching a new patch that (I think) verifies the connection is taken down.

The original test does pass as is, I was surprised about that, too, I was expecting
CancelledError. The docs for UserError do say "user aborted connection", so I guess
it's kind of reasonable.

The new patch fires a CancelledError, but I'm currently setting _suppressAlreadyCalled
to avoid an already called error when the factory gets the connection lost. What's the
right way to proceed there? I guess I need to tell the factory not to fire the deferred.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_3"></a>|@jdavisp3 commented|
|-|-|

Oh, figured it out :)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra removed owner|
|-|-|

Putting it up for review as described on TwistedDevelopment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @jdavisp3|
|-|-|

Thanks.  Some more points, but they're all pretty simple:

  1. The class docstring for `TestQueryFactoryCancel` seems odd.  It doesn't actually record connection lost, but the connector (which keeps track of its own state).
  1. `TestQueryFactoryCancel.test_connector` should be `TestQueryFactoryCancel.testConnector` or perhaps just `TestQueryFactoryCancel.connector`.
  1. The `factory` parameter to all of the `proxy` methods should be documented.
  1. The `test_cancel` docstring shouldn't bother saying _Test that_, as this is evident from the fact that it is a test method. :)
  1. `test_cancel` couold be a little simpler in its handling of `TestQueryFactoryCancel`.  It could instantiate the factory in the `test_cancel` scope, and then just refer to it in the `factory` scope, removing the need for the function attribute.
  1. In the `proxy` method: the `factory or self.queryFactory` idiom is a bit out of favor.  An explicit check for `factory is None` would be better.
  1. Shouldn't all the `proxy` methods use the `factory` argument?  Presently `XMLRPCAllowNoneTestCase.proxy` ignores its.
  1. The `canceller` argument to `_QueryFactory` should be documented.
  1. In the implementation of the cancellation function: errbacking the Deferred with `CancelledError` is actually the behavior of `Deferred.cancel` if the cancellation function doesn't fire or errback the Deferred.  So you can skip the line `d.errback(defer.CancelledError())` in the `cancel` function.  This will actually also provide slightly better behavior because it will lead to the Deferred being errbacked after the connection has been closed, rather than before.  This is nice because it means all resources associated with the Deferred have been cleaned up by the time the Deferred fires.  Without this, there's not really any way to know when the connection is _really_ gone.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_7"></a>|@jdavisp3 commented|
|-|-|

Thanks for the comments. I think I addressed most of them.

Re: #7, since XMLRPCAllowNoneTestCase wasn't using self.queryFactory before, I didn't bother using it. I changed it to not take a factory argument instead.

Re: #5, since the queryFactory is usually instantiated by the Proxy, with arguments gleaned from the URL and the callRemote parameters, it seemed wrong to pre-create it (using hard-coded arguments?).

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

> Re: #5, since the queryFactory is usually instantiated by the Proxy, with arguments gleaned from the URL and the callRemote parameters, it seemed wrong to pre-create it (using hard-coded arguments?). 

Yes, you're right.  I was thinking it'd be a bit cleaner the other way, but now I think I was mistaken.

> Re: #7, since XMLRPCAllowNoneTestCase wasn't using self.queryFactory before, I didn't bother using it. I changed it to not take a factory argument instead. 

Great.

There are a couple other minor things which I'm just going to fix myself, and then I'll commit it:

  * Test methods should be separated from each other by two blank lines.
  * The copyright dates on modified files need to be updated.
  * There should be a news file.
  * When line wrapping text in epytext markup, extra lines should be indented by four spaces, eg:
```
    @foo bar: baz
        quux
```

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_10"></a>|@jdavisp3 commented|
|-|-|

Cool, thanks for the info.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [28819]) Apply patch which makes xmlrpc requests cancellable

Author: davep
Reviewer: exarkun
Fixes: [#4377](https://github.com/twisted/twisted/issues/4377)

Implement a cancellation function for the Deferred returned by
`twisted.web.xmlrpc.Proxy.callRemote`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
